src/plots/rectanglePlot.ts

```diff
@@ -354,6 +400,85 @@ export module Plots {
       });
       return dataToDraw;
     }
+
+    protected _additionalPaint(time: number) {
+      this._renderArea.selectAll(".label-area").remove();
+      if (this._labelsEnabled && this.label() != null) {
+        Utils.Window.setTimeout(() => this._drawLabels(), time);
+      }
+    }
+
+    private _drawLabels() {
+      let dataToDraw = this._getDataToDraw();
+      this.datasets().forEach((dataset, i) => this._drawLabel(dataToDraw, dataset, i));
+    }
+
+    private _drawLabel(dataToDraw: Utils.Map<Dataset, any[]>, dataset: Dataset, datasetIndex: number) {
+      let attrToProjector = this._generateAttrToProjector();
+      let labelArea = this._renderArea.append("g").classed("label-area", true);
+      let measurer = new SVGTypewriter.Measurers.Measurer(labelArea);
+      let writer = new SVGTypewriter.Writers.Writer(measurer);
+      let xRange = this.x().scale.range();
+      let yRange = this.y().scale.range();
+      let xMin = Math.min.apply(null, xRange);
+      let xMax = Math.max.apply(null, xRange);
+      let yMin = Math.min.apply(null, yRange);
+      let yMax = Math.max.apply(null, yRange);
+      let data = dataToDraw.get(dataset);
+      data.forEach((datum, datumIndex) => {
+        let label = "" + this.label()(datum, datumIndex, dataset);
+        let measurement = measurer.measure(label);
+
+        let x = attrToProjector["x"](datum, datumIndex, dataset);
+        let y = attrToProjector["y"](datum, datumIndex, dataset);
+        let width = attrToProjector["width"](datum, datumIndex, dataset);
+        let height = attrToProjector["height"](datum, datumIndex, dataset);
+        if (measurement.height <= height && measurement.width <= width) {
+
+          let horizontalOffset = (width - measurement.width) / 2;
+          let verticalOffset = (height - measurement.height) / 2;
+          x += horizontalOffset;
+          y += verticalOffset;
+
+          let xLabelRange = { min: x, max: x + measurement.width };
+          let yLabelRange = { min : y, max: y + measurement.height };
+          if (xLabelRange.min < xMin || xLabelRange.max > xMax || yLabelRange.min < yMin || yLabelRange.max > yMax) {
+            return;
+          }
+          if (this._overlayLabel(xLabelRange, yLabelRange, datumIndex, datasetIndex, dataToDraw)) {
```

Comment:
Since `_overlayLabel` is an `O(number of points)` call, drawing labels takes `O(n^2)`. This might cause difficulties with larger datasets.